---
title: "RNA Seq Analysis"
output: html_document
---

This code generates the RNA-seq figures (5c and 6a,b) in Stuart et al. 2018

##Preparing the Data
Load in all the libraries.
```{r libraries, warning = FALSE, message = FALSE}
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(RColorBrewer)
library(magrittr)
library(dplyr)
library(scales)
library(readr)
library(BiocParallel)
library(stringr)
```

Load in helper functions
```{r helper functions}
source("./Helper_Functions/dend_sort_exp.R")
source("./Helper_Functions/sort_exp.R")
source("./Helper_Functions/plot_heatmap.R")
source("./Helper_Functions/fancy_scientific.R")
```

Load in the data from HTseq-count output files
```{r data}
directory = "../data"
files = grep("htseq", list.files(directory), value=TRUE)
files[1:9] = files[c(7:9, 4:6, 1:3)]
sampleCondition = factor(c(rep(1,3), rep(2,3), rep(3,3), rep(4,3), rep(5,3), rep(6,3), rep(7,3), rep(8,3), rep(9,2), rep(10,3)))
sample_names = str_remove(files, ".htseq-count.txt")
sampleTable = data.frame(sampleName = sample_names,
                         fileName = files,
                         condition = sampleCondition)
data = DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                  directory = directory,
                                  design = ~ condition)
dim(data)
```

Here is the mapping between number and experimental condition

1. Vector control
2. V600E + 72hr cumate
3. V600E + 72hr cumate + 20h LGX818
4. Week1 Continuous
5. Week2 Continuous
6. Week2 Intermittent **OFF**
7. Week3 Continuous
8. Week3 Intermittent **ON**
9. Week4 Continuous
10. Week4 Intermittent **OFF**


Let's get the gene symbols and names. Currently only have the Ensembl ID which is harder to inspect. Grabbed these names from Ensembl Biomart from the archived Ensembl 84. No filters were selected, and the selected attributes were "Gene Stable ID", "Gene name", "Gene description".
```{r names}
names_file = read.delim("../data/ensembl_gene_name_map.txt", sep = "\t",  quote = "", stringsAsFactors = FALSE)

mapped_names = match(rownames(data), names_file$Ensembl.gene.ID)
gene_names = as.data.frame(names_file[mapped_names,]) #now gene_names is in the same order as the rows of dds
```

#### Supplementary Table 1
Save a csv file of the data for **Table S1**.
```{r table s1}
full_data = data %>% assay %>% as.data.frame
complete_unnormalized_data = bind_cols(gene_names, full_data)
write_csv(complete_unnormalized_data, path = "../results/RNA_Seq_Script_Output/complete_unnormalized_counts.csv") # for Supplementary Table 1
write_delim(complete_unnormalized_data, path = "../results/RNA_Seq_Script_Output/complete_unnormalized_counts.txt", delim = "\t") # for NCBI GEO upload
```

Now we pre-filter the data set to remove any genes that have 0 or 1 read across all conditions. (This reduces memory usage and speeds up the following functions) 

```{r prefilter}
filtered_data = data[ rowSums(counts(data)) > 1, ]
dim(filtered_data)
gene_names = gene_names[ rowSums(counts(data)) > 1, ]
```

There are 60675 genes in the entire dataset, but only 36053 made it through the pre-filtering step. The 60675 genes are from the genome annotation file used (EnsemblGTF file for grch38 version 84).

Now we'll fit the DESeq object and transform the data using the regularized logarithm transform to do visualizations and analyses
```{r deseq object}
dds = DESeq(filtered_data, parallel=TRUE, BPPARAM = MulticoreParam(4))
rld <- rlog(dds, blind=FALSE)
rlog_data = assay(rld)
save(dds, rlog_data, gene_names, file="../data/processed_rna_seq_data.rda")
```

Save tables of filtered raw and normalized counts.
Log fold changes from the mean of vector control cells is in **Supplmentary Table 2** .
```{r more tables}
assay(dds) %>% as.data.frame %>% bind_cols(gene_names) %>% write_csv("../results/RNA_Seq_Script_Output/Filtered_Unnormalized_Data.csv")

norm_counts = (assay(dds) / colData(dds)$sizeFactor) %>% as.data.frame
norm_counts %>% bind_cols(gene_names) %>%
write_csv("../results/RNA_Seq_Script_Output/Filtered_Normalized_Data.csv")

#log fold change from mean of vector control samples
lfc = rlog_data - rowMeans(rlog_data[,c(1:3)])
bind_cols(gene_names, as.data.frame(lfc)) %>% write_csv("../results/RNA_Seq_Script_Output/Supplementary_Table_2.csv") 
```

Let's create a histogram of the distribution of mean and max reads per gene.
```{r histograms}
colnames(norm_counts) = substring(colnames(dds), 3, 5)
gene_means = data.frame(gene_means = rowMeans(norm_counts))
ggplot(gene_means, aes(x = gene_means))+
  geom_freqpoly()+
  ggtitle("Gene Mean")+
  scale_x_log10(breaks = c(0, 0.1, 1, 10, 100, 1000, 10000, 100000),  minor_breaks =c(seq(.2, .9, .1),2:9, seq(20,90,10), seq(200,900,100), seq(2000,9000,1000), seq(20000, 90000, 10000), seq(200000, 900000, 100000)))+
  xlab("Gene Mean")+
  ylab("Count")+
  theme_bw()
ggsave("../results/RNA_Seq_Script_Output/genewise_mean.eps")
gene_max = rowMax(as.matrix(norm_counts))
ggplot(data.frame(gene_max = gene_max), aes(x = gene_max, ..count..))+
  geom_freqpoly()+
  ggtitle("Gene Max")+
  scale_x_log10(labels = fancy_scientific)+
  annotation_logticks()+
  scale_y_continuous()+
  xlab("Gene Max")+
  ylab("Count")+
  theme_bw()
ggsave("../results/RNA_Seq_Script_Output/genewise_max.eps")
```

Let's look at the makeup of the genes in our data. The Ensembl annotation is made up of protein coding genes and other non coding RNAs. Load in these ensembl IDs to look at the distribution. The 
```{r gene types}
gene_types = read.table("../data/ensembl_biomart_gene_type.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

types_map = match(gene_names$Ensembl.gene.ID, gene_types$Ensembl.Gene.ID)
gene_type_summary = data.frame(x = gene_types$Gene.type[types_map])
ggplot(gene_type_summary, aes(x=x))+
  geom_bar()+
  scale_y_continuous("Count", breaks = c(0, 1000, 2500, 5000, 7500, 10000, 12500, 15000,17500), minor_breaks = NULL)+
  scale_x_discrete("Gene Type")+
  theme_bw()+
  ggtitle("Filtered Genes by Type")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
ggsave("../results/RNA_Seq_Script_Output/genes_by_type.eps")
```

## Principal Component Analysis
Perform principal component analysis and plot the first two components
```{r pca}
pca = prcomp(t(rlog_data))
summary(pca)
var_exp = pca$sdev^2/sum(pca$sdev^2) # variance explained
#Make dataframe for ggplot
group <- data.frame(Timepoint = factor(c(rep("Vector", 3), rep("V600E", 3), rep("20hr Drug", 3), rep("Week 1 CONT", 3), rep("Week 2 CONT", 3), rep("Week 2 INT (OFF)", 3), rep("Week 3 CONT", 3), rep("Week 3 INT (ON)", 3), rep("Week 4 CONT", 2), rep("Week 4 INT (OFF)", 3)), levels = c("Vector", "V600E", "20hr Drug", "Week 1 CONT", "Week 2 CONT", "Week 3 CONT", "Week 4 CONT","Week 2 INT (OFF)",  "Week 3 INT (ON)", "Week 4 INT (OFF)")))
dat = data.frame(pc1 =pca$x[,1], pc2 = pca$x[,2], Timepoint = group$Timepoint, names = substr(colnames(dds),1,5))

color_palette = c("#cccccc", "#969696", "#525252", "#ffffcc", "#a1dab4", "#41b6c4", "#225ea8", "#fecc5c",  "#fd8d3c", "#e31a1c")
x_axis = paste("PC1 (", percent(var_exp[1]), " Variance)", sep ="")
y_axis = paste("PC2 (", percent(var_exp[2]), " Variance)", sep ="")
#Now Plot

ggplot(data = dat, aes(x=pc1, y=pc2))+
  geom_point(shape = 21, size = 4, color = "black", stroke = .75, aes(fill = Timepoint))+
  scale_fill_manual(values = color_palette)+
  xlab(x_axis)+
  ylab(y_axis)+
  theme_gray(base_size = 15)+
  theme(text = element_text(family = "Times"))
ggsave("../results/RNA_Seq_Script_Output/PCA_plot.eps", family = "Times")
ggsave("../results/RNA_Seq_Script_Output/PCA_plot.png", height = 4.5, width = 7.3)
```

Get the names of genes with high loadings for each PC
```{r pc genes}
threshold = 0.02
pc1_bool= abs(pca$rotation[,1]) > threshold
pc2_bool = abs(pca$rotation[,2]) > threshold
pc1_genes = which(pc1_bool)
pc2_genes = which(pc2_bool)
pc1_gene_names = gene_names[pc1_genes,]
pc2_gene_names = gene_names[pc2_genes,]

pc1_gene_names %>% bind_cols(., as.data.frame(pca$rotation[pc1_bool, 1])) %>% write_csv(., "../results/RNA_Seq_Script_Output/pc1_genes.csv")
pc2_gene_names %>% bind_cols(., as.data.frame(pca$rotation[pc2_bool, 1])) %>% write_csv(., "../results/RNA_Seq_Script_Output/pc2_genes.csv")
```
### PCA Heatmap
Make a heatmap for PC1 and PC2
```{r pc heatmap}
pc1_only = setdiff(pc1_genes, pc2_genes)
pc2_only = setdiff(pc2_genes, pc1_genes)
both_pc = intersect(pc1_genes, pc2_genes)

lfc_pc1 = rlog_data[pc1_only,] - rowMeans(rlog_data[pc1_only,1:3])
lfc_pc2 = rlog_data[pc2_only,] - rowMeans(rlog_data[pc2_only,1:3])
lfc_both = rlog_data[both_pc,] - rowMeans(rlog_data[both_pc,1:3])

dist.mat = dist(lfc_pc1)
hc1 = hclust(dist.mat, method = "ward.D2")
mean_exp = round(rowMeans(lfc_pc1[hc1$order,c(10:29)]),4)
hc1 = dend_sort_exp(hc1, mean_exp)

dist.mat = dist(lfc_pc2)
hc2 = hclust(dist.mat, method = "ward.D2")
mean_exp = round(rowMeans(lfc_pc2[hc2$order,c(10:29)]),4)
hc2 = dend_sort_exp(hc2, mean_exp)

dist.mat = dist(lfc_both)
hc3 = hclust(dist.mat, method = "ward.D2")
mean_exp = round(rowMeans(lfc_both[hc3$order,c(10:29)]),4)
hc3 = dend_sort_exp(hc3, mean_exp)

plot_pc_data = rbind(lfc_pc1[hc1$order,], lfc_both[hc3$order,], lfc_pc2[hc2$order,])
colnames(plot_pc_data) = c(rep("Vector", 3), rep("V600E", 3), rep("20hr Drug", 3), rep("Week 1 CONT", 3), rep("Week 2 CONT", 3), rep("Week 2 INT (OFF)", 3), rep("Week 3 CONT", 3), rep("Week 3 INT (ON)", 3), rep("Week 4 CONT", 2), rep("Week 4 INT (OFF)", 3))

rownames(plot_pc_data) = c(as.character(gene_names$symbol[pc1_only][hc1$order]), as.character(gene_names$symbol[both_pc][hc3$order]), as.character(gene_names$symbol[pc2_only][hc2$order]))

plot_heatmap(plot_pc_data[, c(1:15, 19:21,25,26,16:18,22:24,27:29)], thresh = 0.97, cluster_rows = FALSE, cluster_cols = FALSE, gaps_row = c(length(pc1_only), length(pc1_only)+length(both_pc)), gaps_col = c(9,20), show_rownames = FALSE, legend_breaks = c(10, 5, 0.97, -0.97, -4, -8))
dev.copy2eps(file="../results/RNA_Seq_Script_Output/heatmap_PCA.eps")
dev.off()
```

## Correlation Plot
Create correlation plot from using the log fold change from the mean of the vector control samples.
```{r corr plot}
lfc = rlog_data - rowMeans(rlog_data[ ,c(1:3)])
cor_mat = cor(lfc)
colnames(cor_mat) = c(rep("Vector Control", 3), rep("V600E Induced", 3), rep("20hr Drug", 3), rep("Week 1", 3), rep("Week 2 CONT",3), rep("Week 2 INT-OFF", 3), rep("Week 3 CONT", 3), rep("Week 3 INT-ON", 3), rep("Week 4 CONT", 2), rep("Week 4 INT-OFF", 3))
rownames(cor_mat) = colnames(cor_mat)
dist.mat_cor = dist(cor_mat)
hc_cor = hclust(dist.mat_cor, method = "ward.D2")
cor_mat = cor_mat[,hc_cor$order]

#color scheme
rc1 = colorRampPalette(colors = c("#00a4ff","#D1E5F0", "#FFFFFF", "#FDDBC7","#ff0000"), space = "Lab")(41)
rb1 = seq(from=-1, to=1, length.out=42)

pheatmap(cor_mat, cluster_rows = TRUE, cluster_cols = FALSE, clustering_method = "ward.D2", color=rc1, breaks=rb1,scale="none", border_color = NA, cellwidth = 9, cellheight = 9, show_colnames = FALSE)
dev.copy2eps(file="../results/RNA_Seq_Script_Output/correlation_matrix.eps")
dev.off()

```

```{r session info}
sessionInfo()
```
